src/renderer/services/plugin-manager.js

```diff
@@ -714,6 +714,67 @@ class PluginManager {
       }
     }
 
+    if (config.permissions.includes('TIMERS')) {
+      const timerArrays = {
+        intervals: [],
+        timeouts: [],
+        timeoutWatchdog: {}
+      }
+
+      const timers = {
+        clearInterval (id) {
+          clearInterval(id)
+          timerArrays.intervals = timerArrays.intervals.filter(interval => interval !== id)
+        },
+
+        clearTimeout (id) {
+          clearTimeout(id)
+          clearTimeout(timerArrays.timeoutWatchdog[id])
+          delete timerArrays.timeoutWatchdog[id]
+          timerArrays.timeouts = timerArrays.timeouts.filter(timeout => timeout !== id)
+        },
+
+        get intervals () {
+          return timerArrays.intervals
+        },
+
+        get timeouts () {
+          return timerArrays.timeouts
+        },
+
+        setInterval (...args) {
+          const id = setInterval(...args)
+          timerArrays.intervals.push(id)
+          return id
+        },
+
+        setTimeout (...args) {
+          const id = setTimeout(...args)
```

Comment:
This needs modifying because the below exposes `require`:

```
    walletApi.timers.setTimeout(function () {
      console.log('setTimeout', this.require)
    }, 100)
```

Probably work better this way:

```
        setTimeout (callback, timer) {
          const id = setTimeout(function () {
            callback()
          }, timer)
          timerArrays.timeouts.push(id)
          timerArrays.timeoutWatchdog[id] = setTimeout(() => timers.clearTimeout(id), timer)
          return id
        }
```

Reply:
Thanks, good catch. Though your proposed approach would break the modern implementation for `setTimeout` and `setInterval` which allows you to use a function as the first argument, timer in ms as the second argument, and further arguments to pass to the called function in subsequent arguments. For example: `setTimeout(console.log, 100, "hello", "world")` should print `hello world` but that functionality would break with the proposed changes.

So I've refactored it like so:

```
setTimeout (...args) {
  const functionArgs = args.splice(2)
  const id = setTimeout(function () {
    args[0](...functionArgs)
  }, args[1])
  timerArrays.timeouts.push(id)
  timerArrays.timeoutWatchdog[id] = setTimeout(() => timers.clearTimeout(id), args[1])
  return id
}
```

Now using `this` doesn't expose anything it shouldn't.